openedx/core/djangoapps/course_groups/cohorts.py

```diff
@@ -222,7 +237,7 @@ def get_cohort(user, course_key, assign=True):
 
     user.course_groups.add(cohort)
 
-    return cohort
+    return request_cache.data.setdefault(cache_key, cohort)
```

Comment:
Probably worth testing to see what the memory growth is like. For other memory related reasons, grading workers are limited to one job before the restart. My guess is that workers just use a single thread for execution (since we spawn entirely separate processes to listen on the queue), but I'm not sure that's the case.

So for now, let's worry about correctness (e.g. there's nothing user-specific in `metadata_inheritance`, right?), and see if memory usage is actually an issue -- both with a simple test grading run of a lot of students and with a back-of-the-napkin calculation of how much we're caching for any given user.